Change non-blocking logger to drop old messages rather than new

cpuguy83 opened the pull request:
Instead of dropping new messages when the buffer is full drop old
messages.
This is more in line with expectations from the system.
The reason for changing the implementation entirely is for performance
This is approx 2x as fast when dropping old messages compared to the
old implementation which just trims the buffer and reslices.
That said, there is a minor performance hit in terms of ns/op (b/c
dropping new messages is always going to be faster), but in the grand
scheme this should not be noticable because this should already be
faster than the configured logger could ever write logs.

The buffer starts with a certain size and will be expanded as needed on
enqueue.
When the queue drops back down below a certain threshold the buffer will
be trimmed back down to prevent excessive/permanant growth.

```
benchmark                                             old ns/op     new ns/op     delta
BenchmarkRingLoggerThroughputNoReceiver-4             15.9          21.5          +35.03%
BenchmarkRingLoggerThroughputWithReceiverDelay0-4     58.1          65.9          +13.36%
BenchmarkRingLoggerThroughputConsumeDelay1-4          16.5          21.5          +30.28%
BenchmarkRingLoggerThroughputConsumeDelay10-4         16.3          21.0          +28.96%
BenchmarkRingLoggerThroughputConsumeDelay50-4         16.4          21.1          +28.72%
BenchmarkRingLoggerThroughputConsumeDelay100-4        16.6          21.2          +27.79%
BenchmarkRingLoggerThroughputConsumeDelay300-4        17.6          20.9          +18.31%
BenchmarkRingLoggerThroughputConsumeDelay500-4        17.1          21.6          +26.63%

benchmark                                             old MB/s     new MB/s     speedup
BenchmarkRingLoggerThroughputNoReceiver-4             1949.19      1444.21      0.74x
BenchmarkRingLoggerThroughputWithReceiverDelay0-4     533.16       470.33       0.88x
BenchmarkRingLoggerThroughputConsumeDelay1-4          1877.76      1441.04      0.77x
BenchmarkRingLoggerThroughputConsumeDelay10-4         1901.50      1474.56      0.78x
BenchmarkRingLoggerThroughputConsumeDelay50-4         1890.40      1468.75      0.78x
BenchmarkRingLoggerThroughputConsumeDelay100-4        1869.00      1462.57      0.78x
BenchmarkRingLoggerThroughputConsumeDelay300-4        1757.37      1485.24      0.85x
BenchmarkRingLoggerThroughputConsumeDelay500-4        1818.58      1435.55      0.79x

benchmark                                             old allocs     new allocs     delta
BenchmarkRingLoggerThroughputNoReceiver-4             0              0              +0.00%
BenchmarkRingLoggerThroughputWithReceiverDelay0-4     0              0              +0.00%
BenchmarkRingLoggerThroughputConsumeDelay1-4          0              0              +0.00%
BenchmarkRingLoggerThroughputConsumeDelay10-4         0              0              +0.00%
BenchmarkRingLoggerThroughputConsumeDelay50-4         0              0              +0.00%
BenchmarkRingLoggerThroughputConsumeDelay100-4        0              0              +0.00%
BenchmarkRingLoggerThroughputConsumeDelay300-4        0              0              +0.00%
BenchmarkRingLoggerThroughputConsumeDelay500-4        0              0              +0.00%

benchmark                                             old bytes     new bytes     delta
BenchmarkRingLoggerThroughputNoReceiver-4             0             0             +0.00%
BenchmarkRingLoggerThroughputWithReceiverDelay0-4     18            6             -66.67%
BenchmarkRingLoggerThroughputConsumeDelay1-4          0             0             +0.00%
BenchmarkRingLoggerThroughputConsumeDelay10-4         0             0             +0.00%
BenchmarkRingLoggerThroughputConsumeDelay50-4         0             0             +0.00%
BenchmarkRingLoggerThroughputConsumeDelay100-4        0             0             +0.00%
BenchmarkRingLoggerThroughputConsumeDelay300-4        0             0             +0.00%
BenchmarkRingLoggerThroughputConsumeDelay500-4        0             0             +0.00%
```

cpuguy83 commented:
Mentioned this in slack, but 2nd commit adds a proper ring buffer implementation that gets within 200MB/sec of the "drop new messages" approach.
I think this is good enough to just switch the implementation over since there's no real trade-off here (other than new messages vs old messages).

PettitWesley commented:
@cpuguy83 I love this new option!! I'm personally against changing the default behavior here though. I recently performed extensive and expensive testing on the AWSLogs driver in non-blocking mode. One thing I found was that in order to understand the performance of non-blocking mode, it must be done statistically, over a very large number of test runs. When I started my testing and ran some simple quick single benchmarks, I got much different (and better looking) results than when I ran them massively in parallel over days. What I found when I ran a massive number of tests was that as log throughput increases, the risk of non-blocking log loss grows. It can be treated like a random phenomenon that happens at some probability I think, and it takes a lot of test runs to measure that probability. This means a single test may easily obtain results that seem very positive.

Please review the results here: https://github.com/moby/moby/issues/45999

cpuguy83 commented:
@PettitWesley Wouldn't the (slightly) lower perf here decrease the chance of dropping a log?

PettitWesley commented:
@cpuguy83 The following is my understanding, please let me know if I'm wrong.
![AWSLogs buffering (3) (2) (1)](https://github.com/moby/moby/assets/29443996/dc1bbd6c-6604-4c7b-9c41-89907af20988)

1. The non-blocking mode buffer is a producer consumer model. The container produces logs into the buffer. The container write call blocks until it returns- until the buffer accepts the logs.
2. The consumer is the log driver, which might be anything, in my case, AWSLogs.
3. Your change slows down the producer side of the equation, the container may take longer to write logs into the buffer when it is full because it goes through longer deletion code. This means the container logging code slows down slightly. 
4. Log loss occurs when the consumer can't keep up with producer.

If the following is correct, then I think there may be no increase in the risk of log loss. The producer will be slowed down slightly occasionally.

cpuguy83 committed:
Change non-blocking log buffer to drop old messages rather than new

Instead of dropping new messages when the buffer is full drop old
messages.
This is more in line with expectations from the system.
The reason for changing the implementation entirely is for performance
This is approx 2x as fast when dropping old messages compared to the
old implementation which just trims the buffer and reslices.
That said, there is a minor performance hit in terms of ns/op (b/c
dropping new messages is always going to be faster), but in the grand
scheme this should not be noticable because this should already be
faster than the configured logger could ever write logs.

The buffer starts with a certain size and will be expanded as needed on
enqueue.
When the queue drops back down below a certain threshold the buffer will
be trimmed back down to prevent excessive/permanant growth.

benchmark                                             old ns/op     new ns/op     delta
BenchmarkRingLoggerThroughputNoReceiver-4             15.9          21.5          +35.03%
BenchmarkRingLoggerThroughputWithReceiverDelay0-4     58.1          65.9          +13.36%
BenchmarkRingLoggerThroughputConsumeDelay1-4          16.5          21.5          +30.28%
BenchmarkRingLoggerThroughputConsumeDelay10-4         16.3          21.0          +28.96%
BenchmarkRingLoggerThroughputConsumeDelay50-4         16.4          21.1          +28.72%
BenchmarkRingLoggerThroughputConsumeDelay100-4        16.6          21.2          +27.79%
BenchmarkRingLoggerThroughputConsumeDelay300-4        17.6          20.9          +18.31%
BenchmarkRingLoggerThroughputConsumeDelay500-4        17.1          21.6          +26.63%

benchmark                                             old MB/s     new MB/s     speedup
BenchmarkRingLoggerThroughputNoReceiver-4             1949.19      1444.21      0.74x
BenchmarkRingLoggerThroughputWithReceiverDelay0-4     533.16       470.33       0.88x
BenchmarkRingLoggerThroughputConsumeDelay1-4          1877.76      1441.04      0.77x
BenchmarkRingLoggerThroughputConsumeDelay10-4         1901.50      1474.56      0.78x
BenchmarkRingLoggerThroughputConsumeDelay50-4         1890.40      1468.75      0.78x
BenchmarkRingLoggerThroughputConsumeDelay100-4        1869.00      1462.57      0.78x
BenchmarkRingLoggerThroughputConsumeDelay300-4        1757.37      1485.24      0.85x
BenchmarkRingLoggerThroughputConsumeDelay500-4        1818.58      1435.55      0.79x

benchmark                                             old allocs     new allocs     delta
BenchmarkRingLoggerThroughputNoReceiver-4             0              0              +0.00%
BenchmarkRingLoggerThroughputWithReceiverDelay0-4     0              0              +0.00%
BenchmarkRingLoggerThroughputConsumeDelay1-4          0              0              +0.00%
BenchmarkRingLoggerThroughputConsumeDelay10-4         0              0              +0.00%
BenchmarkRingLoggerThroughputConsumeDelay50-4         0              0              +0.00%
BenchmarkRingLoggerThroughputConsumeDelay100-4        0              0              +0.00%
BenchmarkRingLoggerThroughputConsumeDelay300-4        0              0              +0.00%
BenchmarkRingLoggerThroughputConsumeDelay500-4        0              0              +0.00%

benchmark                                             old bytes     new bytes     delta
BenchmarkRingLoggerThroughputNoReceiver-4             0             0             +0.00%
BenchmarkRingLoggerThroughputWithReceiverDelay0-4     18            6             -66.67%
BenchmarkRingLoggerThroughputConsumeDelay1-4          0             0             +0.00%
BenchmarkRingLoggerThroughputConsumeDelay10-4         0             0             +0.00%
BenchmarkRingLoggerThroughputConsumeDelay50-4         0             0             +0.00%
BenchmarkRingLoggerThroughputConsumeDelay100-4        0             0             +0.00%
BenchmarkRingLoggerThroughputConsumeDelay300-4        0             0             +0.00%
BenchmarkRingLoggerThroughputConsumeDelay500-4        0             0             +0.00%

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
I've updated this to just switch the internal implementation without providing any new knobs.

cpuguy83 commented:
@corhere mentioned today to check out https://golang.org/pkg/container/ring for this.
Just looking at it, it seems like it would work well. We could even make a copy of it but with generics to get around the `any`.

swapneils commented:
What is the progress on this change? Wasn't this supposed to be merged into docker 25?
